### DISCUSSION TITLE
~trpc.x._output?

### DISCUSSION THREAD

**Discussion by ftzi** [trpc/trpc#3486] [Q&A]
I have a wrapping hook that adds some data and logic to the queries and mutation, to ease their usage in the code. They change the endpoint input by omitting a specific prop that this hook adds.

In trpc@9, with paths as strings, it wasn't that hard to get the queries/mutations inputs and outputs for the generic hooks.

For trpc@10, I managed to do the following to get a query input:

```tsx
type Input<T extends DecorateProcedure<AnyQueryProcedure, any, any>> = Parameters<T['useQuery']>[0]

type A = Input<typeof trpc['x']>
```
But I can't get the output. The following will lead to an `unknown` type:

```tsx
type Output<T extends DecorateProcedure<AnyQueryProcedure, any, any>> = ReturnType<T['useQuery']>

type A = Output<typeof trpc['x']>['data']
```

This is really important to me. Using AppRouter type would be complicated to join its type with trpc.x etc.

My goal is to have useMyQuery(trpc.x.y).

**Comment by ftzi** [comment]
One possible solution would be to have something like this in the type definition. Note that I added it in useQuery and not in the endpoint so it keeps the intellisense clean:

<img width="925" alt="image" src="https://user-images.githubusercontent.com/28177550/209577395-eaf5b277-748c-4b8d-9f87-472a1e4ff868.png">

**Reply by ftzi** [reply]
@KATT would that be a possibility? Would be quite hidden from usual usage and users but still accessible for those who want it. It could also have `_input`, although it's also accessible via `Parameters` type as I showed in the example above.

**Comment by KATT** [comment]
I don't understand the end goal, can you elaborate a bit more there? Are you saying you have an input-prop that you always add on all mutations and queries that you want to abstract into a hook?

Are you saying you want to turn

```ts
const result = trpc.post.byId({
  myUserId: 'x',  // <-- this always sent
  id: 'y'
})

// ...into

const result = myTrpc.post.byId({
  // myUserId: 'x',  // automatically prefilled
  id: 'y'
})
```

Or maybe you're just looking for how to [Infer React Query options based on your router](https://trpc.io/docs/infer-types#infer-react-query-options-based-on-your-router)?

**Reply by ftzi** [reply]
Yes. Most of our queries and mutations have `corpId` as input prop. In trpc@9 we had useQueryWithCorpId, which would prefill it with the user's current corpId value.

**Reply by ftzi** [reply]
(as the user can change corpId, we want them in the query key so we can use the cache and its invalidation with no effort)

**Reply by ftzi** [reply]
Also, not only this, but our hook also sets the query's `enable` based on the `corpId` truthyness (and we also consider `options.enabled`). So, we have some logic besides just defaulting the input value.

**Reply by KATT** [reply]
It's definitely possible, I can do it for you if you can get your employer to pay a few hours of consulting fee or to sign up to be a monthly sponsor at bronze or above

**Reply by ftzi** [reply]
I appreciate but I found another solution for this situation. Still, happy new year!
